voila/handler.py

```diff
@@ -72,26 +68,71 @@ def get(self, path=None):
             config=self.traitlet_config,
             contents_manager=self.contents_manager  # for the image inlining
         )
-
         if self.voila_configuration.strip_sources:
             exporter.exclude_input = True
             exporter.exclude_output_prompt = True
             exporter.exclude_input_prompt = True
 
-        # Filtering out empty cells.
+        cwd = os.path.dirname(notebook_path)
+        # we want to avoid starting multiple kernels due to template mistakes
+        self.kernel_started = False
+
+        # Filter for empty cells.
         def filter_empty_code_cells(cell):
             return (
                 cell.cell_type != 'code' or                     # keep non-code cells
                 (cell.outputs and not exporter.exclude_output)  # keep cell if output not excluded and not empty
                 or not exporter.exclude_input                   # keep cell if input not excluded
             )
-        result.cells = list(filter(filter_empty_code_cells, result.cells))
 
-        html, resources = exporter.from_notebook_node(result, resources=resources)
+        @tornado.gen.coroutine
+        def kernel_start():
+            assert not self.kernel_started
+            # Launch kernel
+            kernel_id = yield tornado.gen.maybe_future(self.kernel_manager.start_kernel(kernel_name=notebook.metadata.kernelspec.name, path=cwd))
+            self.kernel_started = True
+            raise tornado.gen.Return(kernel_id)
+
+        def notebook_execute(nb, kernel_id):
+            km = self.kernel_manager.get_kernel(kernel_id)
+            result = executenb(notebook, km=km, cwd=cwd, config=self.traitlet_config)
+            result.cells = list(filter(filter_empty_code_cells, result.cells))
+            # we modify the notebook in place, since the nb variable cannot be reassigned it seems in jinja2
+            # e.g. if we do {% with nb = notebook_execute(nb, kernel_id) %}, the base template/blocks will not
+            # see the updated variable (it seems to be local to our block)
+            nb.cells = result.cells
+
+        def cell_generator(nb, kernel_id):
+            """Generator that will execute a single notebook cell at a time"""
+            km = self.kernel_manager.get_kernel(kernel_id)
+
+            all_cells = list(nb.cells)  # copy the cells, since we will modify in place
+            for cell in all_cells:
+                # we execute one cell at a time
+                nb.cells = [cell]  # reuse the same notebook
```

Comment:
Does it mean we can't make a single copy of notebook, and use that copy to execute one cell at a time (`nb_copy.cells = [cell]`), while keeping the original notebook intact?

Reply:
we could, but as the code is now, there is no need for it. The general strategy in nbconvert is to do copy.deepcopy. But as of now, each requests reads the notebook file and gives a new notebook object.